File: cmd/cmux_config.go
package cmd

import (
	"encoding/json"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"time"
)

const cmuxConfigPath = ".config/cmux/cmux.json"

var handlerCmuxActions = map[string]map[string]interface{}{
	"handler-switch-to-awaiting": {
		"type":     "command",
		"title":    "agent-handler: Switch to Awaiting Session",
		"subtitle": "Jump to the first session awaiting approval",
		"command":  "handler switch -a --close-caller",
		"shortcut": "cmd+shift+a",
		"palette":  true,
	},
	"handler-switch-to-session": {
		"type":     "command",
		"title":    "agent-handler: Switch to Session",
		"subtitle": "Interactive session switcher with tab completion",
		"command":  "handler switch --close-caller",
		"shortcut": "cmd+shift+s",
		"palette":  true,
	},
	"handler-switch-to-unread": {
		"type":     "command",
		"title":    "agent-handler: Switch to Unread Session",
		"subtitle": "Jump to the first session with unread messages",
		"command":  "handler switch -u --close-caller",
		"shortcut": "cmd+shift+i",
		"palette":  true,
	},
}

var handlerCmuxActionIDs = []string{"handler-switch-to-awaiting", "handler-switch-to-session", "handler-switch-to-unread"}

func cmuxConfigFilePath() string {
	home, _ := os.UserHomeDir()
	return filepath.Join(home, cmuxConfigPath)
}

// readCmuxConfig reads and parses ~/.config/cmux/cmux.json directly. Handler
// owns this end-to-end — it never shells out to any external "cmux-settings"
// helper (that was a skill-provided script that is not guaranteed to be
// installed, and must not be depended on here).
func readCmuxConfig() map[string]interface{} {
	data, err := os.ReadFile(cmuxConfigFilePath())
	if err != nil {
		return nil
	}
	var cfg map[string]interface{}
	if json.Unmarshal(data, &cfg) != nil {
		return nil
	}
	return cfg
}

// writeCmuxConfig backs up the existing cmux.json to a timestamped .bak
// alongside it (per cmux's own docs: "back up any existing cmux.json file to
// a timestamped .bak copy before editing"), then writes the new config and
// asks cmux to reload it.
func writeCmuxConfig(cfg map[string]interface{}) error {
	path := cmuxConfigFilePath()
	if data, err := os.ReadFile(path); err == nil {
		backupPath := fmt.Sprintf("%s.%s.bak", path, time.Now().Format("20060102-150405"))
		if err := os.WriteFile(backupPath, data, 0644); err != nil {
			return fmt.Errorf("backing up cmux.json: %w", err)
		}
	}
	data, err := json.MarshalIndent(cfg, "", "  ")
	if err != nil {
		return fmt.Errorf("marshaling cmux.json: %w", err)
	}
	if err := os.MkdirAll(filepath.Dir(path), 0755); err != nil {
		return fmt.Errorf("creating cmux config dir: %w", err)
	}
	if err := os.WriteFile(path, data, 0644); err != nil {
		return fmt.Errorf("writing cmux.json: %w", err)
	}
	exec.Command("cmux", "reload-config").Run()
	return nil
}

func configureCmuxActions() {
	cfg := readCmuxConfig()
	if cfg == nil {
		cfg = map[string]interface{}{}
	}
	actions, ok := cfg["actions"].(map[string]interface{})
	if !ok {
		actions = map[string]interface{}{}
	}

	// Set each action (always overwrite to pick up updates)
	for _, id := range handlerCmuxActionIDs {
		actions[id] = handlerCmuxActions[id]
	}
	cfg["actions"] = actions

	if err := writeCmuxConfig(cfg); err != nil {
		fmt.Printf("  ⚠ Failed to configure cmux actions: %v\n", err)
		return
	}

	var actionSummary []string
	for _, id := range handlerCmuxActionIDs {
		if s, ok := handlerCmuxActions[id]["shortcut"].(string); ok {
			actionSummary = append(actionSummary, fmt.Sprintf("%s (%s)", id, s))
		}
	}
	fmt.Printf("  ✓ Configured cmux actions: %s\n", strings.Join(actionSummary, ", "))
}

func cmuxConfigActions() map[string]interface{} {
	cfg := readCmuxConfig()
	if cfg == nil {
		return nil
	}
	actions, ok := cfg["actions"].(map[string]interface{})
	if !ok {
		return nil
	}
	return actions
}

func hasCmuxActions() bool {
	actions := cmuxConfigActions()
	if actions == nil {
		return false
	}
	for _, id := range handlerCmuxActionIDs {
		if _, ok := actions[id]; ok {
			return true
		}
	}
	return false
}

// CmuxShortcuts holds the configured keyboard shortcuts for handler cmux actions.
type CmuxShortcuts struct {
	SwitchToAwaiting string
	SwitchToSession  string
	SwitchToUnread   string
	FocusBack        string
	FocusForward     string
}

// GetCmuxShortcuts reads the configured shortcuts directly from
// ~/.config/cmux/cmux.json. Returns nil if the config is missing or the
// handler actions aren't configured in it.
func GetCmuxShortcuts() *CmuxShortcuts {
	cfg := readCmuxConfig()
	if cfg == nil {
		return nil
	}
	actions, _ := cfg["actions"].(map[string]interface{})

	shortcutFor := func(id string) string {
		a, ok := actions[id].(map[string]interface{})
		if !ok {
			return ""
		}
		s, _ := a["shortcut"].(string)
		return s
	}

	shortcuts := &CmuxShortcuts{
		SwitchToAwaiting: shortcutFor("handler-switch-to-awaiting"),
		SwitchToSession:  shortcutFor("handler-switch-to-session"),
		SwitchToUnread:   shortcutFor("handler-switch-to-unread"),
	}

	// Read browser back/forward from cmux shortcuts bindings.
	// These have defaults even without explicit config.
	if shortcutsCfg, ok := cfg["shortcuts"].(map[string]interface{}); ok {
		if bindings, ok := shortcutsCfg["bindings"].(map[string]interface{}); ok {
			if s, ok := bindings["browserBack"].(string); ok && s != "" {
				shortcuts.FocusBack = s
			}
			if s, ok := bindings["browserForward"].(string); ok && s != "" {
				shortcuts.FocusForward = s
			}
		}
	}
	// Default cmux shortcuts if not explicitly configured
	if shortcuts.FocusBack == "" {
		shortcuts.FocusBack = "cmd+["
	}
	if shortcuts.FocusForward == "" {
		shortcuts.FocusForward = "cmd+]"
	}

	if shortcuts.SwitchToAwaiting == "" && shortcuts.SwitchToSession == "" {
		return nil
	}
	return shortcuts
}

func removeCmuxActions() {
	cfg := readCmuxConfig()
	if cfg == nil {
		return
	}
	actions, ok := cfg["actions"].(map[string]interface{})
	if !ok {
		return
	}

	found := false
	for _, id := range handlerCmuxActionIDs {
		if _, ok := actions[id]; ok {
			found = true
			delete(actions, id)
		}
	}
	if !found {
		return
	}
	cfg["actions"] = actions

	if err := writeCmuxConfig(cfg); err != nil {
		fmt.Printf("  ⚠ Failed to remove cmux actions: %v\n", err)
		return
	}
	fmt.Println("  ✓ Removed cmux actions (handler-switch-to-awaiting, handler-switch-to-session)")
}

File: src/Handler/CmuxConfig.hs
-- | Port of cmd/cmux_config.go: handler keyboard-shortcut actions in the
-- cmux config, managed via the cmux-settings helper.
module Handler.CmuxConfig
  ( handlerCmuxActions
  , handlerCmuxActionIDs
  , actionShortcut
  , cmuxConfigFilePath
  , findCmuxSettings
  , configureCmuxActions
  , hasCmuxActions
  , removeCmuxActions
  , CmuxShortcuts(..)
  , getCmuxShortcuts
  ) where

import Control.Monad (forM_, void)
import qualified Data.Aeson as A
import qualified Data.Aeson.Key as Key
import qualified Data.Aeson.KeyMap as KM
import qualified Data.ByteString.Lazy.Char8 as BL
import Data.Maybe (catMaybes, fromMaybe, isJust)
import Data.Text (Text)
import qualified Data.Text as T
import System.Directory (doesFileExist, executable, getHomeDirectory, getPermissions)
import System.Exit (ExitCode(..))
import System.FilePath ((</>))
import System.Process (readProcessWithExitCode)

import Handler.Cli.Common (printfT, putTextLn)

-- | The cmux actions installed by handler setup, keyed by action ID.
handlerCmuxActions :: [(Text, A.Value)]
handlerCmuxActions =
  [ ( "handler-switch-to-awaiting"
    , A.object
        [ "type" A..= ("command" :: Text)
        , "title" A..= ("agent-handler: Switch to Awaiting Session" :: Text)
        , "subtitle" A..= ("Jump to the first session awaiting approval" :: Text)
        , "command" A..= ("handler switch -a --close-caller" :: Text)
        , "shortcut" A..= ("cmd+shift+a" :: Text)
        , "palette" A..= True
        ] )
  , ( "handler-switch-to-session"
    , A.object
        [ "type" A..= ("command" :: Text)
        , "title" A..= ("agent-handler: Switch to Session" :: Text)
        , "subtitle" A..= ("Interactive session switcher with tab completion" :: Text)
        , "command" A..= ("handler switch --close-caller" :: Text)
        , "shortcut" A..= ("cmd+shift+s" :: Text)
        , "palette" A..= True
        ] )
  , ( "handler-switch-to-unread"
    , A.object
        [ "type" A..= ("command" :: Text)
        , "title" A..= ("agent-handler: Switch to Unread Session" :: Text)
        , "subtitle" A..= ("Jump to the first session with unread messages" :: Text)
        , "command" A..= ("handler switch -u --close-caller" :: Text)
        , "shortcut" A..= ("cmd+shift+i" :: Text)
        , "palette" A..= True
        ] )
  ]

handlerCmuxActionIDs :: [Text]
handlerCmuxActionIDs =
  ["handler-switch-to-awaiting", "handler-switch-to-session", "handler-switch-to-unread"]

-- | The \"shortcut\" field of a configured action.
actionShortcut :: Text -> Maybe Text
actionShortcut actionId = do
  A.Object o <- lookup actionId handlerCmuxActions
  A.String s <- KM.lookup "shortcut" o
  pure s

cmuxConfigFilePath :: IO FilePath
cmuxConfigFilePath = do
  home <- getHomeDirectory
  pure (home </> ".config" </> "cmux" </> "cmux.json")

-- | Locates the cmux-settings helper script; Nothing if not installed.
findCmuxSettings :: IO (Maybe FilePath)
findCmuxSettings = do
  home <- getHomeDirectory
  let candidates =
        [ home </> ".agents/skills/cmux-settings/scripts/cmux-settings"
        , home </> ".codex/skills/cmux-settings/scripts/cmux-settings"
        ]
  found <- mapM check candidates
  pure (foldr (\m acc -> if isJust m then m else acc) Nothing found)
  where
    check path = do
      exists <- doesFileExist path
      if not exists
        then pure Nothing
        else do
          perms <- getPermissions path
          pure (if executable perms then Just path else Nothing)

runSettings :: FilePath -> [String] -> IO (Either Text Text)
runSettings helper args = do
  (code, out, errOut) <- readProcessWithExitCode helper args ""
  pure $ case code of
    ExitSuccess -> Right (T.pack out)
    _           -> Left (T.strip (T.pack (out <> errOut)))

-- | Sets each handler action in the cmux config (always overwriting to pick
-- up updates), then reloads cmux config.
configureCmuxActions :: IO ()
configureCmuxActions = do
  findCmuxSettings >>= \case
    Nothing ->
      putTextLn "  \ESC[2mcmux-settings helper not found, skipping cmux action configuration\ESC[0m"
    Just helper -> go helper
  where
    go helper = setAll helper handlerCmuxActionIDs
    setAll helper [] = do
      void $ readProcessWithExitCode "cmux" ["reload-config"] ""
      let summary = T.intercalate ", "
            [ actionId <> " (" <> fromMaybe "" (actionShortcut actionId) <> ")"
            | actionId <- handlerCmuxActionIDs
            ]
      printfT ["  ✓ Configured cmux actions: ", summary]
    setAll helper (actionId : rest) = do
      let actionJson = BL.unpack (A.encode (fromMaybe A.Null (lookup actionId handlerCmuxActions)))
      result <- runSettings helper ["set", "actions." <> T.unpack actionId, actionJson]
      case result of
        Left err -> printfT ["  ⚠ Failed to set cmux action ", actionId, ": ", err]
        Right _  -> setAll helper rest

-- | Whether any handler actions are present in the cmux config.
hasCmuxActions :: IO Bool
hasCmuxActions =
  findCmuxSettings >>= \case
    Nothing -> pure False
    Just helper -> do
      result <- runSettings helper ["get", "actions"]
      pure $ case result of
        Right out | Just (A.Object existing) <- A.decode (BL.pack (T.unpack out)) ->
          any (\actionId -> KM.member (Key.fromText actionId) existing) handlerCmuxActionIDs
        _ -> False

-- | Removes the handler actions from the cmux config and reloads.
removeCmuxActions :: IO ()
removeCmuxActions =
  findCmuxSettings >>= \case
    Nothing -> pure ()
    Just helper -> do
      result <- runSettings helper ["get", "actions"]
      case result of
        Right out | Just (A.Object existing) <- A.decode (BL.pack (T.unpack out))
                  , any (\actionId -> KM.member (Key.fromText actionId) existing) handlerCmuxActionIDs -> do
          forM_ handlerCmuxActionIDs $ \actionId ->
            runSettings helper ["unset", "actions." <> T.unpack actionId]
          void $ readProcessWithExitCode "cmux" ["reload-config"] ""
          putTextLn "  ✓ Removed cmux actions (handler-switch-to-awaiting, handler-switch-to-session)"
        _ -> pure ()

-- | The configured keyboard shortcuts for handler cmux actions.
data CmuxShortcuts = CmuxShortcuts
  { switchToAwaiting :: Text
  , switchToSession  :: Text
  , switchToUnread   :: Text
  , focusBack        :: Text
  , focusForward     :: Text
  } deriving (Show, Eq)

-- | Reads the configured shortcuts from the cmux config.
-- Nothing if cmux-settings is unavailable or actions aren't configured.
getCmuxShortcuts :: IO (Maybe CmuxShortcuts)
getCmuxShortcuts =
  findCmuxSettings >>= \case
    Nothing -> pure Nothing
    Just helper -> do
      result <- runSettings helper ["get", "actions"]
      case result of
        Right out | not (T.null (T.strip out))
                  , Just (A.Object actions) <- A.decode (BL.pack (T.unpack out)) -> do
          let shortcutOf actionId = fromMaybe "" $ do
                A.Object a <- KM.lookup (Key.fromText actionId) actions
                A.String s <- KM.lookup "shortcut" a
                pure s
              awaiting = shortcutOf "handler-switch-to-awaiting"
              session = shortcutOf "handler-switch-to-session"
              unread = shortcutOf "handler-switch-to-unread"
          back0 <- binding helper "shortcuts.bindings.browserBack"
          forward0 <- binding helper "shortcuts.bindings.browserForward"
          -- Default cmux shortcuts if not explicitly configured
          let back = if back0 == "" then "cmd+[" else back0
              forward = if forward0 == "" then "cmd+]" else forward0
          pure $ if awaiting == "" && session == ""
            then Nothing
            else Just CmuxShortcuts
              { switchToAwaiting = awaiting
              , switchToSession = session
              , switchToUnread = unread
              , focusBack = back
              , focusForward = forward
              }
        _ -> pure Nothing
  where
    binding helper key = do
      result <- runSettings helper ["get", key]
      pure $ case result of
        Right out -> T.strip (T.dropAround (== '"') (T.strip out))
        Left _    -> ""
